libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/syncKotlinAndAndroidSourceSets.kt

```diff
@@ -0,0 +1,123 @@
+import org.jetbrains.kotlin.gradle.plugin.addConvention
+import java.io.File
+
+internal fun syncKotlinAndAndroidSourceSets(target: KotlinAndroidTarget) {
+    val project = target.project
+    val android = project.extensions.getByName("android") as BaseExtension
+
+    android.sourceSets.all { androidSourceSet ->
+        val kotlinSourceSetName = kotlinSourceSetNameForAndroidSourceSet(target, androidSourceSet.name)
+        val kotlinSourceSet = project.kotlinExtension.sourceSets.maybeCreate(kotlinSourceSetName)
+        createDefaultDependsOnEdges(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidSourceDirs(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidResources(target, kotlinSourceSet, androidSourceSet)
+        androidSourceSet.addConvention(KOTLIN_DSL_NAME, kotlinSourceSet)
+
+        ifKaptEnabled(project) {
+            Kapt3GradleSubplugin.createAptConfigurationIfNeeded(project, androidSourceSet.name)
+        }
+    }
+}
+
+private fun createDefaultDependsOnEdges(
+    target: KotlinAndroidTarget,
+    kotlinSourceSet: KotlinSourceSet,
+    androidSourceSet: AndroidSourceSet
+) {
+    val commonSourceSetName = when (androidSourceSet.name) {
+        "main" -> "commonMain"
+        "test" -> "commonTest"
+        "androidTest" -> "commonTest"
+        else -> return
+    }
+    val commonSourceSet = target.project.kotlinExtension.sourceSets.findByName(commonSourceSetName) ?: return
+    kotlinSourceSet.dependsOn(commonSourceSet)
+}
+
+private fun syncKotlinAndAndroidSourceDirs(
+    target: KotlinAndroidTarget, kotlinSourceSet: KotlinSourceSet, androidSourceSet: AndroidSourceSet
+) {
+    val disambiguationClassifier = target.disambiguationClassifier
+
+    /*
+    Mitigate ambiguity!
+    Example: disambiguationClassifier="android"
+    Source Directory "src/androidTest/kotlin"
+        -- could be claimed by kotlin {android}Test (unit test)
+        -- could be claimed by Android androidTest (instrumented test)
+
+    The Kotlin source set would win in this scenario.
+     */
+    if (disambiguationClassifier == null || !androidSourceSet.name.startsWith(disambiguationClassifier)) {
+        kotlinSourceSet.kotlin.srcDir("src/${androidSourceSet.name}/kotlin")
+    }
+
+    kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+    androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
+
+    /*
+    Make sure to include user configuration as well.
+    Unfortunately, there does not exist any "ad-hoc" API like `all`.
+    Therefore we sync the directories once again after evaluation
+     */
+    target.project.whenEvaluated {
+        kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+        androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
+    }
+}
+
+private fun syncKotlinAndAndroidResources(
```

Comment:
Can we do this only if MPP is applied?

Reply:
Oh! I see: This *should* not make a difference for non-mpp project because the source sets will have the same name and therefore this paths *should* be the same. I added a check for this case. So if you would (in theory) change the name of the folders, it will only be minorly problematic for mpp projects. 
✅